File: Scripts.lua
--[[--------------------------------------------------------------------
    Copyright (C) 2013, 2014 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--]]--------------------------------------------------------------------

-- This addon is a script repository.

local OVALE, Ovale = ...
local OvaleScripts = Ovale:NewModule("OvaleScripts", "AceEvent-3.0")
Ovale.OvaleScripts = OvaleScripts

--<private-static-properties>
local AceConfig = LibStub("AceConfig-3.0")
local AceConfigDialog = LibStub("AceConfigDialog-3.0")
local OvaleOptions = Ovale.OvaleOptions
local L = Ovale.L

-- Forward declarations for module dependencies.
local OvaleEquipment = nil
local OvalePaperDoll = nil
local OvaleSpellBook = nil
local OvaleStance = nil

local format = string.format
local gsub = string.gsub
local pairs = pairs
local strlower = string.lower

-- Name and description of default script.
local DEFAULT_NAME = "Ovale"
local DEFAULT_DESCRIPTION = L["Script défaut"]

-- Name and description of "custom" script.
local CUSTOM_NAME = "custom"
local CUSTOM_DESCRIPTION = L["Script personnalisé"]

-- Name and description of "disabled" script.
local DISABLED_NAME = "Disabled"
local DISABLED_DESCRIPTION = L["Disabled"]

do
	local defaultDB = {
		code = "",
		source = "Ovale",	-- The name given to all default class scripts.
		showHiddenScripts = false,
	}
	local actions = {
		code  = {
			name = L["Code"],
			type = "execute",
			func = function()
				local appName = OvaleScripts:GetName()
				AceConfigDialog:SetDefaultSize(appName, 700, 550)
				AceConfigDialog:Open(appName)
			end,
		},
	}
	-- Insert defaults into OvaleOptions.
	for k, v in pairs(defaultDB) do
		OvaleOptions.defaultDB.profile[k] = v
	end
	for k, v in pairs(actions) do
		OvaleOptions.options.args.actions.args[k] = v
	end
	OvaleOptions:RegisterOptions(OvaleScripts)
end
--</private-static-properties>

--<public-static-properties>
-- A "script" is a table { type = "scriptType", desc = "description", code = "..." }
-- Table of scripts, indexed by name.
OvaleScripts.script = {}
--</public-static-properties>

--<public-static-methods>
function OvaleScripts:OnInitialize()
	-- Resolve module dependencies.
	OvaleEquipment = Ovale.OvaleEquipment
	OvalePaperDoll = Ovale.OvalePaperDoll
	OvaleSpellBook = Ovale.OvaleSpellBook
	OvaleStance = Ovale.OvaleStance

	self:CreateOptions()
	-- Register the default script that triggers the automatic script selection; the body code is ignored.
	self:RegisterScript(nil, nil, DEFAULT_NAME, DEFAULT_DESCRIPTION, nil, "script")
	-- Register the custom script.
	self:RegisterScript(Ovale.playerClass, nil, CUSTOM_NAME, CUSTOM_DESCRIPTION, Ovale.db.profile.code, "script")
	-- Register an empty script called "Disabled" that can be used to show no icons.
	self:RegisterScript(nil, nil, DISABLED_NAME, DISABLED_DESCRIPTION, nil, "script")
end

function OvaleScripts:OnEnable()
	self:RegisterMessage("Ovale_StanceChanged")
end

function OvaleScripts:OnDisable()
	self:UnregisterMessage("Ovale_StanceChanged")
end

function OvaleScripts:Ovale_StanceChanged(event, newStance, oldStance)
end

-- Return a table of script descriptions indexed by name.
function OvaleScripts:GetDescriptions(scriptType)
	local descriptionsTable = {}
	for name, script in pairs(self.script) do
		if (not scriptType or script.type == scriptType) and (not script.specialization or OvalePaperDoll:IsSpecialization(script.specialization)) then
			if name == DEFAULT_NAME then
				descriptionsTable[name] = script.desc .. " (" .. self:GetScriptName(name) .. ")"
			else
				descriptionsTable[name] = script.desc
			end
		end
	end
	return descriptionsTable
end

function OvaleScripts:RegisterScript(class, specialization, name, description, code, scriptType)
	if not class or class == Ovale.playerClass then
		self.script[name] = self.script[name] or {}
		local script = self.script[name]
		script.type = scriptType or "script"
		script.desc = description or name
		script.specialization = specialization
		script.code = code or ""
	end
end

function OvaleScripts:UnregisterScript(name)
	self.script[name] = nil
end

function OvaleScripts:SetScript(name)
	local oldSource = Ovale.db.profile.source
	if oldSource ~= name then
		Ovale.db.profile.source = name
		self:SendMessage("Ovale_ScriptChanged")
	end
end

function OvaleScripts:GetDefaultScriptName(class, specialization)
	local name
	-- only override when no default script is found in simulationcraft folder
	-- default scripts in the form off "simulationcraft_<class>_<specialization>_t18m"
	if class == "DEATHKNIGHT" then
		if specialization == "blood" then
			name = "icyveins_deathknight_blood"
		elseif specialization == "frost" then
			name = "simulationcraft_death_knight_frost_t19p"
		elseif specialization == "unholy" then
			name = "simulationcraft_death_knight_unholy_t19p"
		end
	elseif class == "DEMONHUNTER" then
		if specialization == "vengeance" then 
			name = "icyveins_demonhunter_vengeance"
		elseif specialization == "havoc" then
			name = "simulationcraft_demon_hunter_havoc_t19p"
		end
	elseif class == "DRUID" then
		if specialization == "restoration" then
			name = DISABLED_NAME
		elseif specialization == "guardian" then
			name = "icyveins_druid_guardian"
		end
	elseif class == "HUNTER" then
		local short
		if specialization == "beast_mastery" then
			short = "bm"
		elseif specialization == "marksmanship" then
			short = "mm"
		elseif specialization == "survival" then
			short = "sv"
		end
		if short then
			name = format("simulationcraft_hunter_%s_t19p", short)
		end
	elseif class == "MONK" then
		if specialization == "mistweaver" then
			name = DISABLED_NAME
		elseif specialization == "brewmaster" then
			name = "icyveins_monk_brewmaster"
		end
	elseif class == "PALADIN" then
		if specialization == "holy" then
			name = DISABLED_NAME
		elseif specialization == "protection" then
			name = "icyveins_paladin_protection"
		end
	elseif class == "PRIEST" then
		if specialization == "discipline" then
			name = "icyveins_priest_discipline"
		elseif specialization == "holy" then
			name = DISABLED_NAME
		end
	elseif class == "SHAMAN" then
		if specialization == "restoration" then
			name = DISABLED_NAME
		end
	elseif class == "WARRIOR" then
		if specialization == "protection" then
			name = "icyveins_warrior_protection"
		end
	end
	if not name and specialization then
		name = format("simulationcraft_%s_%s_t19p", strlower(class), specialization)
	end
	if not (name and self.script[name]) then
		name = DISABLED_NAME
	end
	return name
end

function OvaleScripts:GetScriptName(name)
	return (name == DEFAULT_NAME) and self:GetDefaultScriptName(Ovale.playerClass, OvalePaperDoll:GetSpecialization()) or name
end

function OvaleScripts:GetScript(name)
	name = self:GetScriptName(name)
	if name and self.script[name] then
		return self.script[name].code
	end
end

function OvaleScripts:CreateOptions()
	local options = {
		name = OVALE .. " " .. L["Script"],
		type = "group",
		args = {
			source = {
				order = 10,
				type = "select",
				name = L["Script"],
				width = "double",
				values = function(info)
					local scriptType = not Ovale.db.profile.showHiddenScripts and "script"
					return OvaleScripts:GetDescriptions(scriptType)
				end,
				get = function(info)
					return Ovale.db.profile.source
				end,
				set = function(info, v)
					self:SetScript(v)
				end,
			},
			script = {
				order = 20,
				type = "input",
				multiline = 25,
				name = L["Script"],
				width = "full",
				disabled = function()
					return Ovale.db.profile.source ~= CUSTOM_NAME
				end,
				get = function(info)
					local code = OvaleScripts:GetScript(Ovale.db.profile.source)
					code = code or ""
					-- Substitute spaces for tabs.
					return gsub(code, "\t", "    ")
				end,
				set = function(info, v)
					OvaleScripts:RegisterScript(Ovale.playerClass, nil, CUSTOM_NAME, CUSTOM_DESCRIPTION, v, "script")
					Ovale.db.profile.code = v
					self:SendMessage("Ovale_ScriptChanged")
				end,
			},
			copy = {
				order = 30,
				type = "execute",
				name = L["Copier sur Script personnalisé"],
				disabled = function()
					return Ovale.db.profile.source == CUSTOM_NAME
				end,
				confirm = function()
					return L["Ecraser le Script personnalisé préexistant?"]
				end,
				func = function()
					local code = OvaleScripts:GetScript(Ovale.db.profile.source)
					OvaleScripts:RegisterScript(Ovale.playerClass, nil, CUSTOM_NAME, CUSTOM_DESCRIPTION, code, "script")
					Ovale.db.profile.source = CUSTOM_NAME
					Ovale.db.profile.code = OvaleScripts:GetScript(CUSTOM_NAME)
					self:SendMessage("Ovale_ScriptChanged")
				end,
			},
			showHiddenScripts = {
				order = 40,
				type = "toggle",
				name = L["Show hidden"],
				get = function(info) return Ovale.db.profile.showHiddenScripts end,
				set = function(info, value) Ovale.db.profile.showHiddenScripts = value end
			},
		},
	}

	local appName = self:GetName()
	AceConfig:RegisterOptionsTable(appName, options)
	AceConfigDialog:AddToBlizOptions(appName, L["Script"], OVALE)
end
--</public-static-methods>

File: src/Scripts.ts
import AceConfig from "@wowts/ace_config-3.0";
import AceConfigDialog from "@wowts/ace_config_dialog-3.0";
import { L } from "./Localization";
import { SpecializationName, OvalePaperDollClass } from "./PaperDoll";
import aceEvent, { AceEvent } from "@wowts/ace_event-3.0";
import { format, gsub, lower } from "@wowts/string";
import { pairs, LuaObj, kpairs } from "@wowts/lua";
import { ClassId, SpecializationIndex } from "@wowts/wow-mock";
import { isLuaArray } from "./tools";
import { GetNumSpecializations } from "@wowts/wow-mock";
import { AceModule } from "@wowts/tsaddon";
import { OvaleClass } from "./Ovale";
import { OvaleOptionsClass } from "./Options";

export let DEFAULT_NAME = "Ovale";
let DEFAULT_DESCRIPTION = L["Script défaut"];
const CUSTOM_NAME = "custom";
let CUSTOM_DESCRIPTION = L["Script personnalisé"];
let DISABLED_NAME = "Disabled";
let DISABLED_DESCRIPTION = L["Disabled"];

export type ScriptType = "script" | "include";

interface Script {
    type?: ScriptType;
    desc?: string;
    className?: ClassId;
    specialization?: SpecializationName;
    code?: string;
}

export class OvaleScriptsClass  {

    script:LuaObj<Script> = {}

    private module: AceModule & AceEvent;

    constructor(private ovale: OvaleClass, private ovaleOptions: OvaleOptionsClass, private ovalePaperDoll: OvalePaperDollClass) {
        this.module = ovale.createModule("OvaleScripts", this.OnInitialize, this.OnDisable, aceEvent);

        let defaultDB = {
            code: "",
            source: {},
            showHiddenScripts: false
        }
        let actions = {
            code: {
                name: L["Code"],
                type: "execute",
                func: () => {
                    let appName = this.module.GetName();
                    AceConfigDialog.SetDefaultSize(appName, 700, 550);
                    AceConfigDialog.Open(appName);
                }
            }
        }
        for (const [k, v] of kpairs(defaultDB)) {
            (<any>ovaleOptions.defaultDB.profile)[k] = v;
        }
        for (const [k, v] of pairs(actions)) {
            ovaleOptions.options.args.actions.args[k] = v;
        }
        ovaleOptions.RegisterOptions(this);
    }
    
    private OnInitialize = () => {
        this.CreateOptions();
        this.RegisterScript(undefined, undefined, DEFAULT_NAME, DEFAULT_DESCRIPTION, undefined, "script");
        this.RegisterScript(this.ovale.playerClass, undefined, CUSTOM_NAME, CUSTOM_DESCRIPTION, this.ovaleOptions.db.profile.code, "script");
        this.RegisterScript(undefined, undefined, DISABLED_NAME, DISABLED_DESCRIPTION, undefined, "script");
        this.module.RegisterMessage("Ovale_StanceChanged", this.Ovale_StanceChanged);
        this.module.RegisterMessage("Ovale_ScriptChanged", this.InitScriptProfiles);
    }
    private OnDisable = () => {
        this.module.UnregisterMessage("Ovale_StanceChanged");
        this.module.UnregisterMessage("Ovale_ScriptChanged");
    }
    private Ovale_StanceChanged = (event: string, newStance: string, oldStance: string) => {
    }
    GetDescriptions(scriptType: ScriptType | undefined) {
        let descriptionsTable: LuaObj<string> = {}
        for (const [name, script] of pairs(this.script)) {
            if ((!scriptType || script.type === scriptType) 
            && (!script.className || script.className === this.ovale.playerClass)
            && (!script.specialization || this.ovalePaperDoll.IsSpecialization(script.specialization))) {
                if (name == DEFAULT_NAME) {
                    descriptionsTable[name] = `${script.desc} (${this.GetScriptName(name)})`;
                } else {
                    descriptionsTable[name] = script.desc;
                }
            }
        }
        return descriptionsTable;
    }
    RegisterScript(className: ClassId, specialization: SpecializationName, name: string, description: string, code: string, scriptType: ScriptType) {
        this.script[name] = this.script[name] || {};
        let script = this.script[name];
        script.type = scriptType || "script";
        script.desc = description || name;
        script.specialization = specialization;
        script.code = code || "";
        script.className = className;
    }
    UnregisterScript(name: string) {
        this.script[name] = undefined;
    }
    SetScript(name: string) {
        const oldSource = this.getCurrentSpecScriptName();
        if (oldSource !== name) {
            this.setCurrentSpecScriptName(name);
            this.module.SendMessage("Ovale_ScriptChanged");
        }
    }
    GetDefaultScriptName(className: ClassId, specialization: SpecializationName) {
        let name = undefined;
        let scClassName = lower(className);

        if (className === "DEMONHUNTER") {
            scClassName = "demon_hunter";
        } else if (className === "DEATHKNIGHT") {
            scClassName = "death_knight";
        }

        if (!name && specialization) {
            name = format("sc_t23_%s_%s", scClassName, specialization);
        }
        if (!(name && this.script[name])) {
            name = DISABLED_NAME;
        }
        return name;
    }
    GetScriptName(name: string) {
        return (name == DEFAULT_NAME) && this.GetDefaultScriptName(this.ovale.playerClass, this.ovalePaperDoll.GetSpecialization()) || name;
    }
    GetScript(name: string) {
        name = this.GetScriptName(name);
        if (name && this.script[name]) {
            return this.script[name].code;
        }
    }

    private getCurrentSpecIdentifier() {
        return `${this.ovale.playerClass}_${this.ovalePaperDoll.GetSpecialization()}`;
    }

    getCurrentSpecScriptName() {
        return this.ovaleOptions.db.profile.source[this.getCurrentSpecIdentifier()];
    }

    setCurrentSpecScriptName(source: string) {
        this.ovaleOptions.db.profile.source[this.getCurrentSpecIdentifier()] = source;
    }

    CreateOptions() {
        let options = {
            name: `${this.ovale.GetName()} ${L["Script"]}`,
            type: "group",
            args: {
                source: {
                    order: 10,
                    type: "select",
                    name: L["Script"],
                    width: "double",
                    values: (info: any) => {
                        const scriptType = (!this.ovaleOptions.db.profile.showHiddenScripts && "script") || undefined;
                        return this.GetDescriptions(scriptType);
                    },
                    get: (info: any) => {
                        return this.getCurrentSpecScriptName();
                    },
                    set: (info: any, v: string) => {
                        this.SetScript(v);
                    }
                },
                script: {
                    order: 20,
                    type: "input",
                    multiline: 25,
                    name: L["Script"],
                    width: "full",
                    disabled: () => {
                        return this.getCurrentSpecScriptName() !== CUSTOM_NAME;
                    },
                    get: (info: any)  => {
                        const code = this.GetScript(this.getCurrentSpecScriptName()) || "";
                        return gsub(code, "\t", "    ");
                    },
                    set: (info: any, v: string) => {
                        this.RegisterScript(this.ovale.playerClass, undefined, CUSTOM_NAME, CUSTOM_DESCRIPTION, v, "script");
                        this.ovaleOptions.db.profile.code = v;
                        this.module.SendMessage("Ovale_ScriptChanged");
                    }
                },
                copy: {
                    order: 30,
                    type: "execute",
                    name: L["Copier sur Script personnalisé"],
                    disabled: () => {
                        return this.getCurrentSpecScriptName() === CUSTOM_NAME;
                    },
                    confirm: () => {
                        return L["Ecraser le Script personnalisé préexistant?"];
                    },
                    func: () => {
                        const code = this.GetScript(this.getCurrentSpecScriptName());
                        this.RegisterScript(this.ovale.playerClass, undefined, CUSTOM_NAME, CUSTOM_DESCRIPTION, code, "script");
                        this.setCurrentSpecScriptName(CUSTOM_NAME);
                        this.ovaleOptions.db.profile.code = this.GetScript(CUSTOM_NAME);
                        this.module.SendMessage("Ovale_ScriptChanged");
                    }
                },
                showHiddenScripts: {
                    order: 40,
                    type: "toggle",
                    name: L["Show hidden"],
                    get: (info: any) => {
                        return this.ovaleOptions.db.profile.showHiddenScripts;
                    },
                    set: (info: any, value: boolean) => {
                        this.ovaleOptions.db.profile.showHiddenScripts = value;
                    }
                }
            }
        }
        let appName = this.module.GetName();
        AceConfig.RegisterOptionsTable(appName, options);
        AceConfigDialog.AddToBlizOptions(appName, L["Script"], this.ovale.GetName());
    }

    private InitScriptProfiles = () => {
        let countSpecializations = GetNumSpecializations(false, false);
        if(!isLuaArray(this.ovaleOptions.db.profile.source)){
            this.ovaleOptions.db.profile.source = {}
        }
        for(let i=1; i < countSpecializations; i += 1){
            let specName = this.ovalePaperDoll.GetSpecialization(i as SpecializationIndex)
            this.ovaleOptions.db.profile.source[`${this.ovale.playerClass}_${specName}`] = this.ovaleOptions.db.profile.source[`${this.ovale.playerClass}_${specName}`] || this.GetDefaultScriptName(this.ovale.playerClass, specName);
        }
    }
}
